In [ ]:
import pandas as pd
import numpy as np

labels = ["label_1", "label_2", "label_3", "label_4"]
empty_label = "label_2"
features = [f'feature_{i}' for i in range(1,769)]

Importing data

In [ ]:
train_df = pd.read_csv('train.csv')
train_df.head()

feature_1  feature_2  feature_3  feature_4  feature_5  feature_6  \
0   0.186257  -0.058807   0.024632  -0.163933  -0.146699   0.035889   
1   0.063431  -0.023597   0.068057  -0.252915  -0.061094  -0.027316   
2   0.034962   0.035816  -0.029753  -0.094607  -0.017576  -0.053074   
3   0.033772   0.085612   0.067488  -0.073953  -0.180646  -0.024512   
4   0.134305   0.062096   0.106920  -0.089327   0.117093  -0.077107   

   feature_7  feature_8  feature_9  feature_10  ...  feature_763  feature_764  \
0   0.111708  -0.162861   0.028249   -0.098063  ...     0.055629    -0.010358   
1   0.135747  -0.168147   0.091236   -0.078473  ...    -0.014893     0.071721   
2   0.040121  -0.007932   0.097872   -0.024042  ...     0.012415     0.015215   
3   0.242879  -0.023374  -0.059999    0.002006  ...    -0.078246    -0.032903   
4   0.152579   0.047529  -0.015998   -0.110657  ...    -0.094629     0.069718   

   feature_765  feature_766  feature_767  feature_768  label_1  label_2  \
0     0.125754     0.011648     0.079197     0.093215       45      NaN   
1     0.018918     0.100032    -0.083042     0.088615       45      NaN   
2     0.083808     0.031312    -0.056277     0.064702       45      NaN   
3     0.082949    -0.020659     0.082274    -0.050164       45      NaN   
4     0.014379     0.048124     0.007586    -0.016980       45      NaN   

   label_3  label_4  
0        1        6  
1        1        6  
2        1        6  
3        1        6  
4        1        6  

[5 rows x 772 columns]

In [ ]:
valid_df = pd.read_csv('valid.csv')
valid_df.head()

feature_1  feature_2  feature_3  feature_4  feature_5  feature_6  \
0   0.085129   0.000378  -0.029872  -0.104609  -0.191291  -0.025511   
1   0.222652  -0.094210  -0.056387  -0.210214  -0.294624  -0.032355   
2   0.072147   0.023575  -0.017664  -0.070988  -0.098729   0.018534   
3   0.026623   0.052005  -0.006921  -0.075105  -0.074829  -0.031380   
4   0.112069   0.113702   0.044343   0.010162  -0.169193  -0.099847   

   feature_7  feature_8  feature_9  feature_10  ...  feature_763  feature_764  \
0   0.029381   0.006177   0.079207   -0.012729  ...    -0.026701     0.003306   
1  -0.069078  -0.172146   0.056509   -0.126757  ...    -0.041096     0.013654   
2   0.028282  -0.064789   0.094686   -0.067313  ...    -0.064602    -0.018653   
3   0.012402  -0.043555   0.073825   -0.050979  ...    -0.005655     0.015477   
4   0.195957   0.051861   0.024582    0.001314  ...    -0.118097    -0.043349   

   feature_765  feature_766  feature_767  feature_768  label_1  label_2  \
0     0.052219     0.000784     0.004527     0.102033       45      NaN   
1     0.142623    -0.011882     0.005219     0.143616       45      NaN   
2     0.013636    -0.078994     0.060903     0.064752       45      NaN   
3     0.064209     0.025045    -0.049599     0.078412       45      NaN   
4     0.036754    -0.023065     0.007358    -0.132566       45      NaN   

   label_3  label_4  
0        1        6  
1        1        6  
2        1        6  
3        1        6  
4        1        6  

[5 rows x 772 columns]

In [ ]:
test_df = pd.read_csv('test.csv')
test_df.head()

ID  feature_1  feature_2  feature_3  feature_4  feature_5  feature_6  \
0   1   0.187868  -0.009268   0.039846  -0.102515  -0.109121  -0.079769   
1   2   0.122119   0.019015   0.134483   0.007949  -0.157884  -0.033332   
2   3   0.079187   0.146195  -0.001316  -0.085596  -0.261093  -0.021620   
3   4   0.231196   0.021558   0.215534  -0.210258  -0.158189   0.041621   
4   5   0.020412   0.063732  -0.065506  -0.089598  -0.130788  -0.018809   

   feature_7  feature_8  feature_9  ...  feature_759  feature_760  \
0   0.061295  -0.042890   0.099860  ...     0.043647     0.061174   
1   0.044334  -0.016869  -0.014088  ...    -0.179798    -0.027611   
2   0.268898  -0.103950   0.068976  ...    -0.075460     0.037346   
3   0.240254  -0.068112  -0.017550  ...     0.083334     0.141642   
4   0.119304  -0.057494   0.094714  ...    -0.001969    -0.003025   

   feature_761  feature_762  feature_763  feature_764  feature_765  \
0     0.026771    -0.092734    -0.052383    -0.055526     0.029101   
1     0.240023    -0.005648    -0.057537     0.023922    -0.038126   
2     0.124254    -0.369695    -0.056773    -0.028471     0.019971   
3    -0.007219     0.031547    -0.010762    -0.169036    -0.030963   
4    -0.021156    -0.136993    -0.003615     0.036371     0.048715   

   feature_766  feature_767  feature_768  
0     0.041857    -0.027872     0.099500  
1    -0.015171    -0.006270    -0.003790  
2    -0.065683    -0.047366    -0.121744  
3     0.086698     0.114194    -0.036775  
4    -0.002688    -0.016957    -0.004331  

[5 rows x 769 columns]

In [ ]:
print(train_df.shape)
print(train_df[train_df["label_2"].notna()].shape)

print(valid_df.shape)
print(valid_df[valid_df["label_2"].notna()].shape)

(28520, 772)
(28040, 772)
(750, 772)
(736, 772)


The above result shows that label 2 has empty values in 480 and 14 rows in train and valid rows respectively. They will be handled via deletion.

**Label 2**

In [ ]:
x_train = train_df[train_df['label_2'].notna()]                                              # NaN values are dropped
x_train.drop(labels, axis=1, inplace=True)
y_train = train_df[train_df['label_2'].notna()]['label_2']

x_valid = valid_df[valid_df['label_2'].notna()]                                             # NaN values are dropped
x_valid.drop(labels, axis=1, inplace=True)
y_valid = valid_df[valid_df['label_2'].notna()]['label_2']

x_test = test_df.copy()
x_test.drop(['ID'], axis=1, inplace=True)

<ipython-input-73-0fcb4abba577>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  x_train.drop(labels, axis=1, inplace=True)
<ipython-input-73-0fcb4abba577>:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  x_valid.drop(labels, axis=1, inplace=True)


Feature scaling using robust scaler

In [ ]:
from sklearn.preprocessing import RobustScaler

scaler = RobustScaler()
x_tr_sc = scaler.fit_transform(x_train)
x_vl_sc = scaler.fit_transform(x_valid)
x_test_sc = scaler.fit_transform(x_test)

SVM model

In [ ]:
from sklearn import svm
from sklearn.metrics import accuracy_score

svmc = svm.SVC(kernel="linear")
svmc.fit(x_tr_sc,y_train)

y_pred_svmc = svmc.predict(x_vl_sc)

acc_vl = accuracy_score(y_valid, y_pred_svmc)
acc_vl

0.8125

K-nearest model

In [ ]:
from sklearn.neighbors import KNeighborsClassifier

knnc = KNeighborsClassifier(n_neighbors=50)
knnc.fit(x_tr_sc,y_train)

y_pred_knnc = knnc.predict(x_vl_sc)

acc_vl = accuracy_score(y_valid, y_pred_knnc)
acc_vl

0.7934782608695652

Above results show that SVM classifier has a higher accuracy than k-nearest classifier with deleting of missing values.

Feature engineering with Principal Component Analysis and deletion of missing values

In [ ]:
x_train = train_df[train_df['label_2'].notna()]                                              # NaN values are dropped
x_train.drop(labels, axis=1, inplace=True)
y_train = train_df[train_df['label_2'].notna()]['label_2']

x_valid = valid_df[valid_df['label_2'].notna()]                                             # NaN values are dropped
x_valid.drop(labels, axis=1, inplace=True)
y_valid = valid_df[valid_df['label_2'].notna()]['label_2']

x_test = test_df.copy()
x_test.drop(['ID'], axis=1, inplace=True)

<ipython-input-78-0fcb4abba577>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  x_train.drop(labels, axis=1, inplace=True)
<ipython-input-78-0fcb4abba577>:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  x_valid.drop(labels, axis=1, inplace=True)


In [ ]:
from sklearn.decomposition import PCA

new_feature_numbers = []
accuracy_values = []

n_c = 0.8
while(n_c<0.99):
  pca = PCA(n_components=n_c, svd_solver='full')
  pca.fit(x_train)

  x_train_trf = pd.DataFrame(pca.transform(x_train))
  x_valid_trf = pd.DataFrame(pca.transform(x_valid))
  new_feature_numbers.append(x_train_trf.shape[1])

  classifier = svm.SVC(kernel='linear')
  classifier.fit(x_train_trf,y_train)
  y_pred2 = classifier.predict(x_valid_trf)
  accuracy_values.append(accuracy_score(y_valid,y_pred2))
  n_c=n_c+0.02

print(new_feature_numbers)
print(accuracy_values)

i=0
while(i<len(accuracy_values)):
  print("Accuracy for "+ str(new_feature_numbers[i])+" features :- "+str(accuracy_values[i]))
  i=i+1

[50, 59, 70, 84, 103, 127, 160, 207, 278, 402]
[0.5910326086956522, 0.6005434782608695, 0.623641304347826, 0.657608695652174, 0.6834239130434783, 0.7146739130434783, 0.7432065217391305, 0.7717391304347826, 0.7948369565217391, 0.8247282608695652]
Accuracy for 50 features :- 0.5910326086956522
Accuracy for 59 features :- 0.6005434782608695
Accuracy for 70 features :- 0.623641304347826
Accuracy for 84 features :- 0.657608695652174
Accuracy for 103 features :- 0.6834239130434783
Accuracy for 127 features :- 0.7146739130434783
Accuracy for 160 features :- 0.7432065217391305
Accuracy for 207 features :- 0.7717391304347826
Accuracy for 278 features :- 0.7948369565217391
Accuracy for 402 features :- 0.8247282608695652


Using 0.98 as the n_components value;

In [ ]:
pca = PCA(n_components=0.98, svd_solver='full')
pca.fit(x_train)

x_train_trf = pd.DataFrame(pca.transform(x_train))
x_test_trf = pd.DataFrame(pca.transform(x_test))

classifier = svm.SVC(kernel='linear')
classifier.fit(x_train_trf,y_train)
y_pred = classifier.predict(x_test_trf)
pd.DataFrame(y_pred).to_csv("l2.csv")

In [ ]:
x_test_trf

0         1         2         3         4         5         6    \
0   -0.628112 -0.026265  0.708988 -0.207292 -0.191196 -0.856312  0.655947   
1   -0.738273 -0.217747  1.311791 -0.505747  0.877841  0.223645  0.107033   
2    0.414637  0.833626 -0.522314  1.375745  0.724140 -0.814859  0.566884   
3   -0.378794 -0.431718  0.622446  1.897280  0.058037  1.837101 -0.668329   
4    1.365674  0.080063 -0.110488  0.994157 -0.026177 -0.511805  0.553808   
..        ...       ...       ...       ...       ...       ...       ...   
739  1.418170 -1.621937  0.272770  0.030206 -1.022933 -0.139071  0.103235   
740  0.924644 -0.182993 -0.102863  0.065537  1.080179  0.078615 -0.199548   
741  1.501050 -1.217995  0.253019  0.364966  0.100645  0.398702  0.040593   
742  1.130026 -0.206089  0.116005  0.045699  0.702856  0.272466 -0.435826   
743  0.677118  1.533394  0.401130 -0.361995 -0.595750 -0.281190 -0.833905   

          7         8         9    ...       392       393       394  \
0    0.135774  0.375127 -0.325972  ... -0.005603 -0.012435  0.001217   
1    0.232095  0.235865  0.592040  ...  0.046406  0.001318 -0.021207   
2   -0.356734 -0.079516  0.197574  ...  0.010440 -0.048877 -0.023194   
3    0.788957  0.072010 -0.821842  ...  0.032869  0.025506  0.020303   
4   -0.319103 -0.043312  0.151489  ...  0.001765  0.020901  0.009756   
..        ...       ...       ...  ...       ...       ...       ...   
739 -0.356498  0.336492  0.167020  ...  0.010105  0.009798 -0.014837   
740  0.248502 -0.395931 -0.182330  ... -0.017975 -0.007421 -0.025897   
741  0.252389 -0.059040  0.189346  ... -0.030003  0.044022  0.029924   
742 -0.044024 -0.193530  0.066703  ...  0.035409  0.015665 -0.003747   
743  0.302449 -0.364497  0.278524  ...  0.002771  0.005751  0.018404   

          395       396       397       398       399       400       401  
0   -0.010488 -0.038398 -0.000454 -0.016143  0.005666 -0.004497  0.060123  
1   -0.008533 -0.021879  0.005885  0.031296 -0.023963  0.029626  0.014519  
2    0.020441 -0.004488 -0.008898 -0.040043  0.004963 -0.042820 -0.030372  
3   -0.001002  0.024267 -0.029164  0.007179 -0.024944 -0.027713 -0.021002  
4    0.021580  0.010557  0.019667  0.006661  0.004084  0.000371 -0.009734  
..        ...       ...       ...       ...       ...       ...       ...  
739  0.019466  0.021136 -0.028029  0.053287  0.008957 -0.028619 -0.001450  
740 -0.008687 -0.006094 -0.005626  0.024454 -0.029425  0.034149  0.014407  
741 -0.014059  0.000158  0.008336 -0.004078 -0.000294  0.049147 -0.006109  
742 -0.025480  0.014203  0.009280 -0.012973 -0.015748 -0.026302  0.010521  
743 -0.038156 -0.009114 -0.033636 -0.023313  0.010883 -0.015286  0.031361  

[744 rows x 402 columns]

In [ ]:
train_df = pd.read_csv('train.csv')
valid_df = pd.read_csv('valid.csv')
test_df = pd.read_csv('test.csv')

x_train = train_df.copy()
x_train.drop(labels, axis=1, inplace=True)
y_train_1 = train_df['label_1']
y_train_3 = train_df['label_3']
y_train_4 = train_df['label_4']

x_valid = valid_df.copy()
x_valid.drop(labels, axis=1, inplace=True)
y_valid_1 = valid_df['label_1']
y_valid_3 = valid_df['label_3']
y_valid_4 = valid_df['label_4']

x_test = test_df.copy()
x_test.drop(['ID'], axis=1, inplace=True)

**Label 1**

In [ ]:
new_feature_numbers = []
accuracy_values = []

n_c = 0.8
while(n_c<0.99):
  pca = PCA(n_components=n_c, svd_solver='full')
  pca.fit(x_train)

  x_train_trf = pd.DataFrame(pca.transform(x_train))
  x_valid_trf = pd.DataFrame(pca.transform(x_valid))
  new_feature_numbers.append(x_train_trf.shape[1])

  classifier = svm.SVC(kernel='linear')
  classifier.fit(x_train_trf,y_train_1)
  y_pred2 = classifier.predict(x_valid_trf)
  accuracy_values.append(accuracy_score(y_valid_1,y_pred2))
  n_c=n_c+0.02

print(new_feature_numbers)
print(accuracy_values)

i=0
while(i<len(accuracy_values)):
  print("Accuracy for "+ str(new_feature_numbers[i])+" features :- "+str(accuracy_values[i]))
  i=i+1

[50, 59, 70, 85, 103, 127, 161, 207, 279, 402]
[0.9053333333333333, 0.9213333333333333, 0.9373333333333334, 0.944, 0.9533333333333334, 0.9613333333333334, 0.9613333333333334, 0.9626666666666667, 0.9706666666666667, 0.972]
Accuracy for 50 features :- 0.9053333333333333
Accuracy for 59 features :- 0.9213333333333333
Accuracy for 70 features :- 0.9373333333333334
Accuracy for 85 features :- 0.944
Accuracy for 103 features :- 0.9533333333333334
Accuracy for 127 features :- 0.9613333333333334
Accuracy for 161 features :- 0.9613333333333334
Accuracy for 207 features :- 0.9626666666666667
Accuracy for 279 features :- 0.9706666666666667
Accuracy for 402 features :- 0.972


Using 0.98 as the n_components value;

In [ ]:
pca = PCA(n_components=0.98, svd_solver='full')
pca.fit(x_train)

x_train_trf = pd.DataFrame(pca.transform(x_train))
x_test_trf = pd.DataFrame(pca.transform(x_test))

classifier = svm.SVC(kernel='linear')
classifier.fit(x_train_trf,y_train_1)
y_pred = classifier.predict(x_test_trf)
pd.DataFrame(y_pred).to_csv("l1.csv")

**Label 3**

In [ ]:
new_feature_numbers = []
accuracy_values = []

n_c = 0.8
while(n_c<0.99):
  pca = PCA(n_components=n_c, svd_solver='full')
  pca.fit(x_train)

  x_train_trf = pd.DataFrame(pca.transform(x_train))
  x_valid_trf = pd.DataFrame(pca.transform(x_valid))
  new_feature_numbers.append(x_train_trf.shape[1])

  classifier = svm.SVC(kernel='linear')
  classifier.fit(x_train_trf,y_train_3)
  y_pred2 = classifier.predict(x_valid_trf)
  accuracy_values.append(accuracy_score(y_valid_3,y_pred2))
  n_c=n_c+0.02

print(new_feature_numbers)
print(accuracy_values)

i=0
while(i<len(accuracy_values)):
  print("Accuracy for "+ str(new_feature_numbers[i])+" features :- "+str(accuracy_values[i]))
  i=i+1

[50, 59, 70, 85, 103, 127, 161, 207, 279, 402]
[0.9906666666666667, 0.9946666666666667, 0.9933333333333333, 0.996, 0.9946666666666667, 0.9986666666666667, 0.9986666666666667, 0.9986666666666667, 0.9986666666666667, 0.9986666666666667]
Accuracy for 50 features :- 0.9906666666666667
Accuracy for 59 features :- 0.9946666666666667
Accuracy for 70 features :- 0.9933333333333333
Accuracy for 85 features :- 0.996
Accuracy for 103 features :- 0.9946666666666667
Accuracy for 127 features :- 0.9986666666666667
Accuracy for 161 features :- 0.9986666666666667
Accuracy for 207 features :- 0.9986666666666667
Accuracy for 279 features :- 0.9986666666666667
Accuracy for 402 features :- 0.9986666666666667


Since accuracy stays constant after *n_components* becomes 0.9, that value, with 127 features are chosen.

In [ ]:
pca = PCA(n_components=0.9, svd_solver='full')
pca.fit(x_train)

x_train_trf = pd.DataFrame(pca.transform(x_train))
x_test_trf = pd.DataFrame(pca.transform(x_test))

classifier = svm.SVC(kernel='linear')
classifier.fit(x_train_trf,y_train_3)
y_pred = classifier.predict(x_test_trf)
pd.DataFrame(y_pred).to_csv("l3.csv")

**Label 4**

In [ ]:
train_df['label_4'].value_counts()

6     19938
2      1449
0       955
12      954
7       938
13      482
1       481
11      480
10      480
3       479
5       478
9       472
4       469
8       465
Name: label_4, dtype: int64

Since the distribution is uneven, an additional parameter, *class_weight* with a value 'balanced' is added to the svc classifier.

In [ ]:
new_feature_numbers = []
accuracy_values = []

n_c = 0.8
while(n_c<0.99):
  pca = PCA(n_components=n_c, svd_solver='full')
  pca.fit(x_train)

  x_train_trf = pd.DataFrame(pca.transform(x_train))
  x_valid_trf = pd.DataFrame(pca.transform(x_valid))
  new_feature_numbers.append(x_train_trf.shape[1])

  classifier = svm.SVC(kernel='linear', class_weight="balanced")
  classifier.fit(x_train_trf,y_train_4)
  y_pred2 = classifier.predict(x_valid_trf)
  accuracy_values.append(accuracy_score(y_valid_4,y_pred2))
  n_c=n_c+0.02

print(new_feature_numbers)
print(accuracy_values)

i=0
while(i<len(accuracy_values)):
  print("Accuracy for "+ str(new_feature_numbers[i])+" features :- "+str(accuracy_values[i]))
  i=i+1

[50, 59, 70, 85, 103, 127, 161, 207, 279, 402]
[0.5773333333333334, 0.6266666666666667, 0.644, 0.7, 0.7333333333333333, 0.7666666666666667, 0.8, 0.8373333333333334, 0.8586666666666667, 0.8786666666666667]
Accuracy for 50 features :- 0.5773333333333334
Accuracy for 59 features :- 0.6266666666666667
Accuracy for 70 features :- 0.644
Accuracy for 85 features :- 0.7
Accuracy for 103 features :- 0.7333333333333333
Accuracy for 127 features :- 0.7666666666666667
Accuracy for 161 features :- 0.8
Accuracy for 207 features :- 0.8373333333333334
Accuracy for 279 features :- 0.8586666666666667
Accuracy for 402 features :- 0.8786666666666667


Using 0.98 as the n_components value;

In [ ]:
pca = PCA(n_components=0.98, svd_solver='full')
pca.fit(x_train)

x_train_trf = pd.DataFrame(pca.transform(x_train))
x_test_trf = pd.DataFrame(pca.transform(x_test))

classifier = svm.SVC(kernel='linear')
classifier.fit(x_train_trf,y_train_4)
y_pred = classifier.predict(x_test_trf)
pd.DataFrame(y_pred).to_csv("l4.csv")